### DISCUSSION TITLE
Multiple instruments

### DISCUSSION THREAD

**Discussion by alenpavlovich** [kernc/backtesting.py#418] [Ideas]
Hello backtesting team,

Does the library support multiple instruments?

To be specific:

- load data for each instrument
- trade any instrument
- have a position on any instrument 
- have an indicator on any instruments 
- have PnL on each instrument

**Comment by alenpavlovich** [comment]
Any thoughts on the following? The reason I am asking is if you wanna test the long-short strategy. Or trade futures and the spot.

**Reply by alenpavlovich** [reply]
Bump

**Reply by atakanokan** [reply]
another bump from me

**Comment by AgarwalPragy** [comment]
@kernc I would like to take this up. 

I'm having some trouble figuring out how exactly the classes should be modified to handle the same.
Any help would be appreciated 😄 

-- --

My current idea is to accept `data` as either a `pd.DataFrame` or a dictionary containing instrument names (`str`) and corresponding instrument data (`pd.DataFrame`)

All classes will internally store data always as a dictionary (if there is a single instrument, it will be named `default` and stored in the dict).

For `data.index`, if there is only 1 instrument, the behavior remains as it is. If there are multiple instruments, we can force the index to be `pd.DateTimeIndex` and then union all the indices. Note that with multiple instruments, it is possible that some data values are `NaN` since newer companies might not have data going back too far, causing different instruments might have different starting dates.

Update:

A better idea would be to update the class `_Data` to always work with multiple instruments. All the internal code will assume multiple instruments. However, the class `Backtest` will allow the user to pass a single dataframe to maintain backwards compatibility.

**Reply by AgarwalPragy** [reply]
Ah.. looks like someone is already working on it https://github.com/kernc/backtesting.py/pull/639

**Reply by AgarwalPragy** [reply]
I've made a parallel attempt here https://github.com/kernc/backtesting.py/pull/641
